Enhance configuration documentation (#348)
ethyca/fides: docs/fides/docs/installation/configuration.md

By default the fidesctl CLI doesn't require a config file and will instead leverage the default values. These are very likely to be wrong however so it is recommended to always configure your settings properly.

## Configuration file

After initializing fidesctl, a default configuration file will be generated and placed within the `.fides` directory. Here's an example of a default fidesctl configuration file and documentation for each section:

=== "default fidesctl.toml"

    ```toml
    [api]
    database_url = "postgres:fidesctl@fidesctl-db:5432/fidesctl"

    # Logging
    log_destination = ""
    log_level = INFO
    log_serialization = ""

    [cli]
    local_mode = False
    server_url = "http://localhost:8080"

    [user]
    encryption_key = "test_encryption_key"

>     ```

Reviewer: This doesn't feel logically connected to the config option tables, so it should probably just live inline. This will also help readers compare the values in the tables to the values here.

=== "API Section"

    | Name | Type | Default | Description |
    | :----: | :----: | :-------: | :-----------: |

>     | test_database_url | String | ""| If the `FIDESCTL_TEST_MODE` environment variable is set to `True`, the `test_database_url` is used instead of the `database_url` to avoid overwriting production data. |
>     | database_url | String | postgres:fidesctl@fidesctl-db:5432/fidesctl | The PostgreSQL database connection string for the fidesctl database. __NOTE__: Do not include the driver here, fidesctl will do this for you. |

Reviewer: Let's swap these lines so it reads a little more intuitively.

```suggestion
    | database_url | String | `postgres:fidesctl@fidesctl-db:5432/fidesctl` | The PostgreSQL database connection string for the fidesctl database. **Do not** include a database driver here; fidesctl will include it automatically. |
    | test_database_url | String | ""| Used instead of the `database_url` when the `FIDESCTL_TEST_MODE` environment variable is set to `True`, to avoid overwriting production data. |
```

Author: done

>     | log_destination | String | "" | The output location for log files. Accepts any valid file path. If left unset, no log files are produced. |

Reviewer: ```suggestion
    | log_destination | String | "" | The output location for log files. Accepts any valid file path. If left unset, log entries are printed to `stdout` and log files are not produced. |
```

Author: done

    | log_level | Enum (String) | INFO | The minimum log entry level to produce. Also accepts: `TRACE`, `DEBUG`, `WARNING`, `ERROR`, or `CRITICAL` (case insensitive). |
    | log_serialization | Enum (String) | "" | The format with which to produce log entries. If left unset, produces log entries formatted using the internal custom formatter. Also accepts: `"JSON"` (case insensitive). |

=== "CLI Section"

    | Name | Type | Default | Description |
    | :----: | :----: | :-------: | :-----------: |

>     | local_mode | Boolean | False | When to `True`, the CLI will never attempt to call out to a fidesctl webserver. |
>     | server_url | String | "" | The URL for the fidesctl webserver that the CLI should connect to. |

Reviewer: ```suggestion
    | local_mode | Boolean | False | When set to `True`, forbids the fidesctl CLI from making calls to the fidesctl webserver. |
    | server_url | String | "" | The base URL of the fidesctl webserver endpoints, in `host:port` format. |
```

=== "User Section"

    | Name | Type | Default | Description |
    | :----: | :----: | :-------: | :-----------: |

>     | encryption_key | String | "" | The key used to encrypt the user's data stored in the database. |

Reviewer: I break my own rule about leaking implementation details, but in this case I think it's valuable information. Without specifying the cryptography algorithm, users might think they're required to provide an AES256-compatible bytestring, for example.
```suggestion
    | encryption_key | String | "" | An arbitrary string used to encrypt the user data stored in the database. Encryption is implemented using PGP. |
```

By default the fidesctl CLI doesn't require a config file and will instead leverage the default values. These are very likely to be wrong however so it is recommended to always configure your settings properly.

## Configuration file

After initializing fidesctl, a default configuration file will be generated and placed within the `.fides` directory. Here's an example of a default fidesctl configuration file:

```toml title="fidesctl.toml"
[api]
database_url = "postgres:fidesctl@fidesctl-db:5432/fidesctl"

# Logging
log_destination = ""
log_level = INFO
log_serialization = ""

[cli]
local_mode = False
server_url = "http://localhost:8080"

[user]
encryption_key = "test_encryption_key"
```

To better describe the various configuration options, the following tables describe each available option, grouped by section:

=== "API Section"

    | Name | Type | Default | Description |
    | :----: | :----: | :-------: | :-----------: |
    | database_url | String | postgres:fidesctl@fidesctl-db:5432/fidesctl | The PostgreSQL database connection string for the fidesctl database. __NOTE__: Do not include the driver here, fidesctl will do this for you. |

>     | test_database_url | String | ""| If the `FIDESCTL_TEST_MODE` environment variable is set to `True`, the `test_database_url` is used instead of the `database_url` to avoid overwriting production data. |

Reviewer: ```suggestion
    | test_database_url | String | ""| Used instead of the `database_url` when the `FIDESCTL_TEST_MODE` environment variable is set to `True`, to avoid overwriting production data. |
```

```toml title="fidesctl.toml"
[api]

# If FIDESCTL_TEST_MODE env var is set to True, the test_database_url
# will be used instead of the database URL to avoid overwriting production data
test_database_url = "postgres:fidesctl@fidesctl-db:5432/fidesctl_test"

> # The SQLAlchemy connection string used to connect to the database
> # https://docs.sqlalchemy.org/en/14/core/engines.html#postgresql

Reviewer: I don't believe anything about the connection string is specific to SQLAlchemy, and naming SQLAlchemy here + linking to the docs leaks implementation details.
```suggestion
# The PostgreSQL database connection string for the fidesctl database
```

Author: regardless of the tool, sqlalchemy has a very useful page on connection strings. what is the negative of leaking implementation details?

Reviewer: In general it's a bad practice. It opens up potential attack vectors, couples the documentation to the implementation, and may confuse inexperienced users who could be led to believe that these are somehow SQLAlchemy-specific. I don't believe we need to guide users on building connection strings as part of our docs (similar to [my below comment](https://github.com/ethyca/fides/pull/348#discussion_r801037507), I think we should decompose the config into it the connection's components and build the string ourselves), but if you'd like to link to something, why not [the pSQL docs](https://www.postgresql.org/docs/14/libpq-connect.html#LIBPQ-CONNSTRING) instead?

Author: added

database_url = "postgres:fidesctl@fidesctl-db:5432/fidesctl"

# Logging
log_destination = "" # Also accepts: Any valid file path
log_level = INFO # Also accepts: TRACE, DEBUG, WARNING, ERROR, CRITICAL

> log_serialization = "" # Also accepts: JSON

Reviewer: I don't know if it's worth mentioning, but the values here are not case sensitive. Including them in documentation in all-caps implies [to me] that they are.

Author: added

[cli]

> local_mode = False # Tells fidesctl to run without calling a webserver

Reviewer: I'm struggling with a suggestion here, but I think setting this to `False` and then saying that the option "tells fidesctl to run without calling a webserver" is slightly confusing - it implies that the `False` setting is what achieves this.
```suggestion
local_mode = True # Run fidesctl without any external connections (default: False)
```

Author: added a better description, good call

> server_url = "http://localhost:8080" # The URL of the fidesctl webserver

Reviewer: Probably work for another PR, but how would you feel about decomposing this into `server_host` and `server_port`, then building the URL internally where it's needed? It would better align our config with other tools I've used, and reduce things like typos, expectations of HTTPs, etc.

Reviewer: Done - created #355 !

[user]

> # The secure key used to encrypt sensitive information in the database

Reviewer: This feels vague, but I'm not familiar enough with the requirements here to suggest something more helpful to a new user.

Author: @SteveDMurphy could you give a more concise description here?

also, what happens if no key is provided? will the values from the db just show up as encrypted or will it fail?

Reviewer: I spoke to @SteveDMurphy and dug into the code a bit. I think we should 1) add some validation for presence of an encryption key, and 2) change our default to literally any string. Regardless, using `""` would not break anything; it's used as a PGP key under the hood.
